### PR TITLE
Fix accessing editor theme items throughout the UI

### LINKED ISSUE #81492: Some editor icons are broken in the ui (master branch)
### Godot version

Godot v4.2.dev (fc99492d3)

### System information

Godot v4.2.dev (fc99492d3) - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 970 (NVIDIA; 31.0.15.3623) - Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz (8 Threads)

### Issue description

Some editor icons are broken and show a red placeholder icon in current master at fc99492d3. Looks like a recent regression, 4.2 dev4 shows these icons correctly.

### Steps to reproduce

Icons that I have noticed that are affected, there can be more:

* Path3D gizmo handle icons
* MeshInstance3D toolbar button icon
* VoxelGI toolbar button icon
* LightmapGI toolbar button icon

icons

mesh

### Minimal reproduction project

Not needed.

### LINKED ISSUE #81586: Theme color for 'error_color' is no longer red
### Godot version

v4.2.dev.custom_build [3ed449711]

### System information

Godot v4.2.dev (3ed449711) - macOS 13.5.1 - Vulkan (Forward+) - integrated Apple M1 - Apple M1 (8 Threads)

### Issue description

The error output is no longer red. The color is assigned with (0,0,0, 1) (black)
https://github.com/godotengine/godot/blob/master/editor/export/project_export.cpp#L1416

Bildschirmfoto 2023-09-12 um 21 21 56

**Comparison with the version v4.2.dev4.official [549fcce5f]**

Bildschirmfoto 2023-09-12 um 21 34 28

### Steps to reproduce

Force error output in Project->Export dialog (like no templates or delete "Unique Name" in Android template)

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by YuriSizov**
Closes https://github.com/godotengine/godot/issues/81492 (there may be other cases, but what's reported should be fixed now). Closes https://github.com/godotengine/godot/issues/81586.

### TL;DR

This PR:

- Fixes accessing theme items through `EditorInterface::get_base_control`/`EditorNode::get_gui_base`, which no longer works in some contexts.
- Fixes accessing some non-existent editor theme items.
- Adds `EditorIterface::get_editor_theme` to be used by user plugins and by our own C# plugin.
- Adds a warning system for theme misses in the editor theme, which helps a bit but not a lot.

### How we got here

https://github.com/godotengine/godot/pull/81130 changed some things about when and how it would be appropriate to access the editor theme. Previously, we would set it directly on the GUI base node (as well as on a dedicated theme base node, for some reason, and on the editor's main window). While not incorrect in principle, this setup meant that the editor theme would propagate to each and every control or window node as long as the chain of themable nodes was uninterrupted. This led to theme leaks.

As the goal of https://github.com/godotengine/godot/pull/81130 was to fix these leaks, among other things, the approach had to be changed. Now we define theme contexts for the editor UI. This means that the GUI base node no longer holds a theme. Instead, the editor theme is used through a global reference. Unfortunately, throughout our own codebase we would often rely on the `get_base_control()->get_theme_*` pattern to access icons and whatnot. And this did work before — even during the editor initialization, even before the editor UI was ready. It worked because the node would hold the theme and would be able to provide these theme properties at any given moment.

We even recommended this approach in the documentation to assign editor icons as user plugin icons.

This is pretty bad, because this relies on something that is not guaranteed to be stable. It was fine for the time being, but a GUI rework made it break, as you can see. Preserving compatibility here would be detrimental to our ability to develop the editor. Adding the theme back to the GUI base control would reintroduce the leaks and would also slow down the editor startup.

It's a huge shame that we suggested this approach to users, though this was probably the only alternative at the time. We should've exposed a method to provide editor icons directly, without relying on GUI nodes and their state. Specifically so these icons could be used outside of the GUI code, such as in

```gdscript

func _get_plugin_icon():
    return EditorInterface.get_base_control().get_theme_icon("Node", "EditorIcons")
```

### How I'm addressing this

For our internal code it's solvable with some find and replace, by accessing `EditorNode::get_singleton()->get_editor_theme`. This resource is ready early on, and can be used by most of the editor code. Some parts of the codebase that rely on this access are GUI elements and should actually be rewritten as dedicated GUI components, so that they can benefit from the `NOTIFICATION_THEME_CHANGED` signal and correctly update when users change the editor theme. But I left it as is for now, with some TODOs.

Gizmos are a bit trickier, as these icons are used in constructors to pre-populate materials. I haven't worked with gizmos for quite some time, let alone with Godot 4 gizmos, so I don't remember if we could recreate these materials later on, or if it would be a bad idea. For now I left them as is. But it may be sometime to look into (also may want to add some helper methods so that creating materials from editor icons doesn't take over a hundred characters per line).

Then come user plugins, and our own C# editor plugin, which is limited, for the most part, by our public API. For these I'm adding a new method to `EditorInterface`, which allows fetching the editor theme. I'm not too happy about adding more methods to `EditorInterface` when my goal is to split its responsibilities between multiple single-purpose classes, but that's the only place available now, so alas.

For the C# editor plugin I recommend reworking it a bit, so the editor run bar icon is correctly updated with the editor theme. Currently it's only set when the plugin is enabled, and ignores theme changes. We've fixed a similar issue in the build panel before. I should've identified this when reviewing https://github.com/godotengine/godot/pull/79357. cc @raulsntos 

For user plugins https://github.com/godotengine/godot/pull/81130 kind of becomes a breaking change, as far as accessing editor icons goes. This doesn't affect plugins that use custom icons, and this doesn't affect GUI code in these plugins, as long is it's written correctly (using theme propagation). Things like `_get_plugin_icon()` and inopportune accesses to `get_base_control` would need to be fixed by plugin creators. The new `EditorInterface::get_editor_theme` method will prove useful here, and it should also be possible to update plugins in a way that works both with 4.1 and 4.2, if that's required.

I think that's an acceptable compromise.

### Reporting theme misses

@Calinou suggested:

> Can we make it so requesting an unknown icon in the editor always prints an error at least once? This would make it easier to track down those issues in the future.

I like the idea, but I'm not sure there is a good way to implement this without adding editor-specific code to controls and windows, and I don't like that idea. The reason why this approach may be required is that most of these issues are because of trying to access icons and styles before there is the editor theme, or before you can access it in some context. So this can only be checked by the node itself.

For now I've added a new type, `EditorTheme` which provides warnings for cases when the editor theme is there, but the theme item requested is wrong. This actually caught a few issues, but none of the ones reported in https://github.com/godotengine/godot/issues/81492. I fixed them together with others.

The new `EditorTheme` class is added as a separate independent commit and can be dropped or reverted later. There are some implementation details that I would like to change so it's less of a nuisance to maintain, but that can be done later. Trying to keep changes to the minimum here.

### CODE REVIEW

File: editor/editor_node.cpp [hunk 1]
```diff
@@ -7153,15 +7151,6 @@ EditorNode::EditorNode() {
...
 	main_menu->set_menu_tooltip(0, TTR("Operations with scene files."));
 
-	prev_scene = memnew(Button);
-	prev_scene->set_flat(true);
-	prev_scene->set_tooltip_text(TTR("Go to previously opened scene."));
-	prev_scene->set_disabled(true);
-	prev_scene->connect("pressed", callable_mp(this, &EditorNode::_menu_option).bind(FILE_OPEN_PREV));
-	gui_base->add_child(prev_scene);
-	prev_scene->set_position(Point2(3, 24));
-	prev_scene->hide();
```
Comment (YuriSizov) [inline editor/editor_node.cpp:7163]: Some old control, but it doesn't appear to be used in any way. It's always hidden, and it's... just floating? I noticed it before, but didn't look further. But now it's removed because it's also requesting a non-existent icon.

File: editor/editor_interface.h [hunk 2]
```diff
@@ -84,11 +85,10 @@ class EditorInterface : public Object {
 	void set_plugin_enabled(const String &p_plugin, bool p_enabled);
 	bool is_plugin_enabled(const String &p_plugin) const;
 
-	void add_editor_plugin(EditorPlugin *p_plugin);
-	void remove_editor_plugin(EditorPlugin *p_plugin);
-
```
Comment (YuriSizov) [inline editor/editor_interface.h:89]: These aren't implemented on `EditorInterface`.

File: editor/plugins/theme_editor_preview.cpp [hunk 3]
```diff
@@ -217,7 +217,7 @@ void ThemeEditorPreview::_notification(int p_what) {
 			theme_cache.preview_picker_overlay_color = get_theme_color(SNAME("preview_picker_overlay_color"), SNAME("ThemeEditor"));
 			theme_cache.preview_picker_label = get_theme_stylebox(SNAME("preview_picker_label"), SNAME("ThemeEditor"));
 			theme_cache.preview_picker_font = get_theme_font(SNAME("status_source"), EditorStringName(EditorFonts));
-			theme_cache.font_size = get_theme_font_size(SNAME("font_size"), EditorStringName(EditorFonts));
+			theme_cache.font_size = get_theme_default_font_size();
```
Comment (YuriSizov) [inline editor/plugins/theme_editor_preview.cpp:220]: This was added in https://github.com/godotengine/godot/pull/54221, but there is no `font_size` in `EditorFonts`. I assume the intent here is to get some fallback value, hence my suggested change. cc @bruvzg

Comment (lmurray) [reply editor/plugins/theme_editor_preview.cpp:220]: Since the font size value is always used with `preview_picker_font` which is derived from `status_source`, maybe the `status_source_size` font size is the intended value to use here?

Comment (YuriSizov) [reply editor/plugins/theme_editor_preview.cpp:220]: I see. Well, in that case I think that my suggested change should be fine for now. Just because I use `status_source` font for the label in the picker mode, it doesn't mean I have to use its corresponding font size. The default one fits the design of the tool, so I'll leave it as is.

Thanks for feedback!

**Comment by YuriSizov** [comment]
I noticed a couple more issues, one was caught by the new check (so, yay!).

- Some recent changes to the visual shaders plugin made a typo and were trying to fetch fonts from the icons type.
- The create node dialog was missing icons because it was requesting them through another node for some reason, and that node didn't have the necessary information yet.

**Comment by YuriSizov** [comment]
I went ahead and replaced all remaining `get_gui_base()->get_theme_*` with `get_editor_theme()->get_*`. This addresses issues like https://github.com/godotengine/godot/issues/81586. In a lot of cases relevant code should likely be reworked to utilize the theme changed notification and correctly set colors and styles when it's received.

**Comment by KoBeWi** [comment]
Some remaining broken icon in GridMap:
image

File: doc/classes/EditorInterface.xml [hunk 4]
```diff
@@ -109,6 +109,13 @@
 				Returns the editor's [EditorSettings] instance.
 			</description>
 		</method>
+		<method name="get_editor_theme" qualifiers="const">
+			<return type="Theme" />
+			<description>
+				Returns the editor's [Theme].
+				[b]Note:[/b] When creating custom editor UI prefer accessing theme items directly from your GUI nodes using [code]get_theme_*[/code] methods.
```
Comment (AThousandShips) [inline doc/classes/EditorInterface.xml:116]:
```suggestion
				[b]Note:[/b] When creating custom editor UI, prefer accessing theme items directly from your GUI nodes using the [code]get_theme_*[/code] methods.
```
Long sentence

**Comment by YuriSizov** [comment]
> Some remaining broken icon in GridMap: image

Right, forgot to include the new `get_editor_theme_icon` calls when replacing. Adjusted those as well, mostly with a search and replace though in a couple of places I tried to correct the code to avoid calling into the editor node all together.

File: editor/editor_themes.h [hunk 5]
```diff
@@ -53,6 +53,22 @@ class EditorColorMap {
 	static HashSet<StringName> &get_color_conversion_exceptions() { return color_conversion_exceptions; };
 };
 
+class EditorTheme : public Theme {
+	GDCLASS(EditorTheme, Theme);
+
+	static Vector<StringName> editor_theme_types;
```
Comment (KoBeWi) [inline editor/editor_themes.h:59]: Why is this static if it's not used in a static context? Also there is only one EditorTheme.

File: editor/plugins/lightmap_gi_editor_plugin.cpp [hunk 6]
```diff
@@ -168,7 +169,9 @@ void LightmapGIEditorPlugin::_bind_methods() {
 LightmapGIEditorPlugin::LightmapGIEditorPlugin() {
 	bake = memnew(Button);
 	bake->set_flat(true);
-	bake->set_icon(EditorNode::get_singleton()->get_gui_base()->get_editor_theme_icon(SNAME("Bake")));
+	// TODO: Rework this as a dedicated toolbar control so we can hook into theme changes and update it
```
Comment (KoBeWi) [inline editor/plugins/lightmap_gi_editor_plugin.cpp:172]: Why did you add this comment here if this plugin is the same as the other plugins? (e.g. GPUParticlesCollisionSDF3DEditorPlugin)

File: editor/editor_themes.h (hunk 5 above)
Comment (YuriSizov) [reply editor/editor_themes.h:59]: Well, if for some reason there is more than one EditorTheme instanced at the same time (even if transiently), they don't need to duplicate this list which is tied to the editor as a whole.

Basically, I made it static because it's not tied to any instance.

File: editor/plugins/lightmap_gi_editor_plugin.cpp (hunk 6 above)
Comment (YuriSizov) [reply editor/plugins/lightmap_gi_editor_plugin.cpp:172]: IIRC when I fixed that part I only had two affected plugins in front of me, and I added the same comment to both. Others came with one of the force pushes where it was mostly a blind change, and I didn't think about adding more comments.

Comment (YuriSizov) [reply editor/plugins/lightmap_gi_editor_plugin.cpp:172]: I can add similar comments all over the place, if that's a concern :) But I think if someone stumbles upon this, they would check if there are other cases (or we would nudge them towards them during the review). If I don't get to these first, that is.

Comment (KoBeWi) [reply editor/plugins/lightmap_gi_editor_plugin.cpp:172]: > But I think if someone stumbles upon this, they would check if there are other cases

Such comments would make them easier to find, but I guess you'll be the one to fix it anyway.

**Review by KoBeWi — APPROVED** [review]

File: editor/editor_themes.cpp [hunk 7]
```diff
@@ -214,6 +215,103 @@ void EditorColorMap::create() {
...
+
+// TODO: Refactor these and corresponding Theme methods to use the bool get_xxx(r_value) pattern internally.
+
+// Keep in sync with Theme::get_color.
+Color EditorTheme::get_color(const StringName &p_name, const StringName &p_theme_type) const {
+	if (color_map.has(p_theme_type) && color_map[p_theme_type].has(p_name)) {
+		return color_map[p_theme_type][p_name];
+	} else {
+		if (editor_theme_types.has(p_theme_type)) {
+			WARN_PRINT(vformat("Trying to access a non-existing editor theme color '%s' in '%s'", p_name, p_theme_type));
```
Comment (AThousandShips) [inline editor/editor_themes.cpp:228]:
```suggestion
			WARN_PRINT(vformat("Trying to access a non-existing editor theme color '%s' in '%s'.", p_name, p_theme_type));
```
Same for the rest

**Comment by YuriSizov** [comment]
Thanks for reviews!
